types/dec_coin.go

```diff
@@ -333,6 +333,22 @@ func (coins DecCoins) Intersect(coinsB DecCoins) DecCoins {
 	return removeZeroDecCoins(res)
 }
 
+// Min returns the minimum of each denom of its inputs, including
+// the implicitly-zero absent denoms. Identical to Intersect() and
+// provided for compatibility with Coins.
```

Comment:
Not including but excluding, right?

Reply:
The missing denom is _included_ in the calculation of minimums, but _excluded_ from the result because it's zero. I was trying to draw a contrast with your earlier interpretation. I'd be open to better wording if you have any to suggest.

Reply:
The methods in this PR operate on the union set of two sets of coins, on a per-denom basis. Therefore, intuitively, as long as a denom exists in one of those input sets, it should also be represented in the output set.

If a denom has to be present in both sets in order to be represented in the output set, then say exactly that in the docs 👍

Reply:
I think the sparse-set implementation of `Coins` is causing confusion here, since any `Coins` value includes quantities for all denominations, since `AmountOf()` always returns a number instead of occasional errors. Would this phrasing be better?

```
// Min returns a Coins value that contains, for each denom, the minimum of the amount in either of its inputs.
```

Reply:
That sounds correct to me. AmountOf will return 0 if the denom isn't present.

I guess if we wanted to be very explicitly, we could add a Note saying that the length of the returned object can be less than the length of both inputs. I don't think such a note is necessary, but I am also quite normalized to the sparse representation of the coins object.

Reply:
> // Min returns a Coins value that contains, for each denom, the minimum of the amount in either of its inputs.

Given two sets of coins `{D1:100}` and `{}` — the minimum amount of D1 is 100, not 0.

Reply:
That's the crux of the issue, isn't it? I'll get some more senior folk to weigh in on the common understanding, and see what I can do to make the behavior plain as day to everyone's understanding, common or not.

For what it's worth, when I see `Coins{}` I also see `NewCoins(Coin("D1", ZeroInt()), Coin("D2", ZeroInt()), ...)`, since the latter evaluates to the former.

But @peterbourgon , aside from the documentation, do you have a use case for the alternative semantics that you suggested? Or were you just describing the behavior you thought was implied by the documentation comment? If there's another valid use case, we could tweak the naming to make the different semantics more clear.

Reply:
> `Coins{}` [evaluates to] `NewCoins(Coin("D1", ZeroInt()), Coin("D2", ZeroInt()), ...)`

That's very surprising to me. The Coins type represents a set, right? Assuming so: an empty set is definitely not equivalent to a set containing an arbitrary number of elements with value=0. Said another way, if something doesn't exist, that's not the same thing as it's value being zero. 

```
Min(x, {x:0} {x:3}) = 0
Min(x, {x:2} {   }) = 2
Min(x, {   } {   }) = undefined
```

But this is all probably just noise. If the Coins type asserts otherwise, then sure, yolo or whatever 😉

Reply:
> A [multiset](https://en.wikipedia.org/wiki/Multiset) of denom strings can be equivalently envisioned as a set that allows repeats ({"atom", "atom", "atom", "eth"}) 

✔︎ 

> or as a set where each element has a nonzero multiplicity ({("atom", 3), ("eth", 1)})

✔︎ 

> or as a mapping from the entire space of values to nonnegatives ({("atom", 3), ("bld", 0), ("btc", 0), ..., ("eth", 1), ("hex": 0), ...}).

✘ 

The entire set of possible denoms is infinite and unknowable, which would make Coins an [infinite set](https://en.wikipedia.org/wiki/Infinite_set). This is nonintuitive, and produces surprising results for common set operations — like Min! :) I'm not saying operations which treat Coins as infinite sets shouldn't exist, just that (a) this shouldn't be the default assumption, and (b) those operations should be unambiguously named and documented regarding this assumption and its consequences.

Reply:
In the alternative you object to, I explicitly said "mapping", not "implementation". To be fair, I should have written `"atom" -> 3, "bld" -> 0, "btc" -> 0, ..., "eth" -> 1, "hex" -> 0, ...` to make it super clear that I'm describing a mapping. This is exactly the mulitiplicity function from the cited article, and is implemented (in finite space) by `AmountOf()`. Despite the infinite domain of possible denom strings, all these examples are finite multisets and have finite cardinality.  Specifying the mapping `"btc" -> 0` does not increase the cardinality.

Every other `Coins` operation works with the semantics implemented for `Min()` - the result is, for every one of the infinite denoms, the equivalent scalar operation for that denom:

```
AmountOf(denom, a.Add(b)) == a.AmountOf(denom).Add(b.AmountOf(denom))
AmountOf(denom, a.Sub(b)) == a.AmountOf(denom).Sub(b.AmountOf(denom))
```

Reply:
> Every other `Coins` operation works with the semantics implemented for `Min()`

And I'd also say that e.g. [v0.45.1 Coins.Add](https://pkg.go.dev/github.com/cosmos/cosmos-sdk/types#Coins.Add) is equally confusing! `{2A} Add {A, 2B}` should intuitively yield `{3A 2B}` and not `{3A}`. But I'm clearly not being convincing here so I defer 👍

Reply:
I guess there's lots of surprising properties about Coin/Coins. A Coin with a zero amount should definitely be valid. A zero value Coins, or equivalently a Coins without any Coins in it, should definitely be valid. Negative amounts should be representable. And so on. So maybe that's the windmill I'm tilting against here, really. In any case, I defer 👍

Reply:
I think it is just an optimization: we don't store zero balances in the state (to free the state), nor we don't store zero coins in messages, because they don't change any value.

Reply:
If you're talking about a zero quantity for a denom in a `Coins` value, then I'd put it even more strongly - you can't possibly represent them explicitly since there are too many possible denom strings.

If you're talking about the all-zeroes `Coins` value, then it doesn't make sense for some operations, e.g. a send of empty `Coins` between accounts seems odd. On the other hand, in Issue 11223 I gave a long list of places where an all-zeroes `Coins` value is essential.